Source/Core/VideoCommon/VertexLoaderManager.cpp

```diff
@@ -198,56 +197,42 @@ NativeVertexFormat* GetUberVertexFormat(const PortableVertexDeclaration& decl)
   return GetOrCreateMatchingFormat(new_decl);
 }
 
-static VertexLoaderBase* RefreshLoader(int vtx_attr_group, bool preprocess = false)
+template <bool IsPreprocess>
+VertexLoaderBase* GetOrCreateLoader(int vtx_attr_group)
```

Comment:
This function could be passed in the `BitSet8` and the vertex loader, since the if check was already done in `RefreshLoader` .

Reply:
Does that matter? IMO it doesn't improve the code and hopefully the compiler will just optimize the ternaries away.

Reply:
I wouldn't see why the ternarys would be optimized away, they are a branch.  Maybe the compiler would condense the 3 branches into one though.  I'm not sure.

It's not a huge gain or anything, just figured I'd mention it, it's easy.

Reply:
The condition for the ternary is known at compile time, since it's a template argument.

Reply:
It seems like it's legal to write this, which would make it explicit that it's known at compile time:

```C++

  constexpr CPState* state = IsPreprocess ? &g_preprocess_cp_state : &g_main_cp_state;
  constexpr BitSet8& attr_dirty = IsPreprocess ? g_preprocess_vat_dirty : g_main_vat_dirty;
  constexpr auto& vertex_loaders =
      IsPreprocess ? g_preprocess_vertex_loaders : g_main_vertex_loaders;
```

Reply:
Hmm.  But I thought that's the point of the `if constexpr` ?  It's to avoid the branch because it's known at compile time.

Reply:
I believe that the compiler is still allowed to optimize out if-statements that it always knows the path of, but `if constexpr` is allowed to have things that would fail to compile in regular `if` statements in the path that is not evaluated. At least, that's what the examples given [on cppreference](https://en.cppreference.com/w/cpp/language/if#Constexpr_if) show.

Reply:
> I believe that the compiler is still allowed to optimize out if-statements that it always knows the path of

Absolutely. As long as the program still behaves the same, the compiler can do whatever it wants, and this is one of those cases.

Reply:
Interesting.  I did not know the compiler would make those sorts of optimizations, I guess it makes sense.  I might try and write up a godbolt to see.  Appreciate the info!

@Pokechu22 - my understanding was that `constexpr` on a variable/method gave no guarantee that the variable/method would be deduced at compile-time, just that you're asking really nicely :).  Still, this would make me feel a bit better if we don't want to jump through more hoops.

Reply:
It doesn't guarantee that, but on the other hand, it does reject this:

```C++
VertexLoaderBase* GetOrCreateLoader(int vtx_attr_group, bool is_preprocess)
{
  constexpr CPState* state = is_preprocess ? &g_preprocess_cp_state : &g_main_cp_state;
  constexpr BitSet8& attr_dirty = is_preprocess ? g_preprocess_vat_dirty : g_main_vat_dirty;
  constexpr auto& vertex_loaders =
      is_preprocess ? g_preprocess_vertex_loaders : g_main_vertex_loaders;
  // ...
}
```

with "expression did not evaluate to a constant: the value of parameter `is_preprocess` cannot be used as a constant"

Reply:
Confirmed with a godbolt.  Thanks all for the info!

Reply:
Added `constexpr` to the ternary for good measure.